File: RESTAPITestSwaggerDoc/Energy-api-tests/src/test/java/com/example/EnergyApiTests.java
package com.example;

import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.response.Response;
import org.junit.BeforeClass;
import org.junit.Test;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import static io.restassured.RestAssured.*;
import static org.hamcrest.Matchers.*;

public class EnergyApiTests {

    private static final String BASE_URL = "https://qacandidatetest.ensek.io/ENSEK";
    private static final int DEFAULT_QUANTITY = 10;
    private static List<Integer> orderIds = new ArrayList<>();
    private static String firstOrderDate;

    @BeforeClass
    public static void setup() {
        RestAssured.baseURI = BASE_URL;
    }

    // Helper: Buy energy
    private int buyEnergy(int energyId, int quantity) {
        Response response = given()
                .pathParam("id", energyId)
                .pathParam("quantity", quantity)
                .when()
                .put("/buy/{id}/{quantity}")
                .then()
                .statusCode(200)
                .extract().response();

        return response.jsonPath().getInt("orderId");
    }

    // Helper: Get energy IDs
    private List<Integer> getEnergyIds() {
        Response response = given()
                .when()
                .get("/api/energy-ids")
                .then()
                .statusCode(200)
                .extract().response();

        List<Integer> ids = response.jsonPath().getList("id");
        if (ids == null || ids.isEmpty()) throw new IllegalStateException("No energy IDs retrieved.");
        return ids;
    }

    // Helper: Verify order details
    private void verifyOrderDetails(int orderId, int quantity) {
        given().pathParam("orderId", orderId)
                .when().get("/orders/{orderId}")
                .then().statusCode(200)
                .body("id", equalTo(orderId))
                .body("quantity", equalTo(quantity));
    }

    @Test
    public void testResetDataUnauthorized() {
        given().when().post("/reset")
                .then().statusCode(401);
    }

    @Test
    public void testBuyEnergyAndStoreOrders() {
        for (int id : getEnergyIds()) {
            int orderId = buyEnergy(id, DEFAULT_QUANTITY);
            orderIds.add(orderId);
        }
    }

    @Test
    public void testGetOrdersAndVerifyDetails() {
        Response ordersResponse = given()
                .when().get("/orders")
                .then().statusCode(200)
                .extract().response();

        List<String> creationDates = ordersResponse.jsonPath().getList("creationDate");
        if (!creationDates.isEmpty()) firstOrderDate = creationDates.get(0);

        for (int orderId : orderIds) {
            verifyOrderDetails(orderId, DEFAULT_QUANTITY);
        }
    }

    @Test
    public void testCountOrdersBeforeNow() {
        if (firstOrderDate == null) throw new IllegalStateException("No order date available.");

        List<String> dates = given()
                .when().get("/orders")
                .then().statusCode(200)
                .extract().jsonPath().getList("creationDate");

        long count = dates.stream()
                .filter(d -> d.compareTo(LocalDateTime.now().toString()) < 0)
                .count();

        System.out.println("Orders before current time: " + count);
    }

    @Test
    public void testUnauthorizedLogin() {
        String payload = "{\"username\":\"wrongUser\",\"password\":\"wrongPassword\"}";
        given().contentType(ContentType.JSON).body(payload)
                .when().post("/login")
                .then().statusCode(401)
                .body("message", equalTo("Unauthorized"));
    }

    @Test
    public void testBadRequestOnBuy() {
        given().pathParam("id", 1)
                .pathParam("quantity", -5)
                .when().put("/buy/{id}/{quantity}")
                .then().statusCode(400)
                .body("message", equalTo("Bad Request"));
    }
}

File: EnsekSeleniumTest/EnergyApiTests.cs
using Xunit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using RestSharp;

namespace EnsekSeleniumTest
{
    public class EnergyApiTests
    {
        private const string LoginUrl = "https://qacandidatetest.ensek.io/ENSEK/login";
        private const string EnergyUrl = "https://qacandidatetest.ensek.io/ENSEK/energy";
        private const string BuyUrl = "https://qacandidatetest.ensek.io/ENSEK/buy";
        private const string ResetUrl = "https://qacandidatetest.ensek.io/ENSEK/reset";
        private const string OrdersUrl = "https://qacandidatetest.ensek.io/ENSEK/orders";

        private const int DefaultQuantity = 10;
        private static readonly List<string> OrderIds = new();
        private static string? FirstOrderTime;
        private readonly RestClient client;
        private string? accessToken;

        public EnergyApiTests()
        {
            client = new RestClient();
        }

        private string Authenticate()
        {
            if (!string.IsNullOrEmpty(accessToken)) return accessToken;

            var request = new RestRequest(LoginUrl, Method.Post);
            request.AddHeader("Content-Type", "application/json");
            request.AddStringBody("{\"username\":\"test\",\"password\":\"testing\"}", DataFormat.Json);

            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            accessToken = json.RootElement.GetProperty("access_token").GetString();
            return accessToken!;
        }

        private RestRequest CreateRequest(string url, Method method, bool auth = false)
        {
            var request = new RestRequest(url, method);
            if (auth)
                request.AddHeader("Authorization", $"Bearer {Authenticate()}");
            return request;
        }

        private string BuyEnergy(int energyId, int quantity)
        {
            var request = CreateRequest(BuyUrl, Method.Post, auth: true);
            request.AddHeader("Content-Type", "application/json");
            request.AddStringBody($"{{\"energy_id\":{energyId},\"quantity\":{quantity}}}", DataFormat.Json);

            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            return json.RootElement.GetProperty("id").GetString()!;
        }

        private List<int> GetEnergyIds()
        {
            var request = CreateRequest(EnergyUrl, Method.Get);
            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            var ids = new List<int>();
            foreach (var prop in json.RootElement.EnumerateObject())
            {
                ids.Add(prop.Value.GetProperty("energy_id").GetInt32());
            }
            return ids;
        }

        private void VerifyOrderDetails(string orderId, int quantity)
        {
            var request = CreateRequest($"{OrdersUrl}/{orderId}", Method.Get, auth: true);
            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            Assert.Equal(orderId, json.RootElement.GetProperty("id").GetString());
            Assert.Equal(quantity, json.RootElement.GetProperty("quantity").GetInt32());
        }

        [Fact]
        public void TestUnauthorizedLogin()
        {
            var request = new RestRequest(LoginUrl, Method.Post);
            request.AddHeader("Content-Type", "application/json");
            request.AddStringBody("{\"username\":\"wrongUser\",\"password\":\"wrongPassword\"}", DataFormat.Json);

            var response = client.Execute(request);
            Assert.Equal(401, (int)response.StatusCode);
        }

        [Fact]
        public void TestResetDataAuthorized()
        {
            var request = CreateRequest(ResetUrl, Method.Post, auth: true);
            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);
        }

        [Fact]
        public void TestBuyEnergyAndStoreOrders()
        {
            foreach (var id in GetEnergyIds())
            {
                var orderId = BuyEnergy(id, DefaultQuantity);
                OrderIds.Add(orderId);
            }
        }

        [Fact]
        public void TestGetOrdersAndVerifyDetails()
        {
            if (!OrderIds.Any())
                TestBuyEnergyAndStoreOrders();

            var request = CreateRequest(OrdersUrl, Method.Get, auth: true);
            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            var orders = json.RootElement.EnumerateArray().ToList();
            if (orders.Count > 0)
                FirstOrderTime = orders[0].GetProperty("time").GetString();

            foreach (var orderId in OrderIds)
                VerifyOrderDetails(orderId, DefaultQuantity);
        }

        [Fact]
        public void TestCountOrdersBeforeNow()
        {
            if (FirstOrderTime == null)
                TestGetOrdersAndVerifyDetails();

            var request = CreateRequest(OrdersUrl, Method.Get, auth: true);
            var response = client.Execute(request);
            Assert.Equal(200, (int)response.StatusCode);

            using var json = JsonDocument.Parse(response.Content!);
            var times = json.RootElement.EnumerateArray()
                .Select(x => x.TryGetProperty("time", out var t) ? t.GetString()! : string.Empty)
                .Where(s => !string.IsNullOrEmpty(s))
                .Select(t => DateTime.Parse(t))
                .ToList();

            var count = times.Count(d => d < DateTime.UtcNow);
            Console.WriteLine($"Orders before current time: {count}");
        }

        [Fact]
        public void TestBadRequestOnBuy()
        {
            var request = CreateRequest(BuyUrl, Method.Post, auth: true);
            request.AddHeader("Content-Type", "application/json");
            request.AddStringBody("{\"energy_id\":1,\"quantity\":-5}", DataFormat.Json);

            var response = client.Execute(request);
            Assert.Equal(400, (int)response.StatusCode);
        }
    }
}
